clients/vault/src/oracle/agent.rs

```diff
@@ -57,84 +87,104 @@ async fn handle_message(
 	Ok(())
 }
 
-/// Start the connection to the Stellar Node.
-/// Returns an `OracleAgent` that will handle incoming messages from Stellar Node,
-/// and to send messages to Stellar Node
-pub async fn start_oracle_agent(
+pub async fn listen_for_stellar_messages(
 	config: StellarOverlayConfig,
-	secret_key: &str,
+	oracle_agent: ArcRwLock<OracleAgent>,
+	secret_key_as_string: String,
 	shutdown_sender: ShutdownSender,
-) -> Result<OracleAgent, Error> {
-	let is_public_network = config.is_public_network();
+) -> Result<(), service::Error<crate::Error>> {
+	tracing::info!(
+		"listen_for_stellar_messages(): Starting connection to Stellar overlay network..."
+	);
 
-	tracing::info!("start_oracle_agent(): Starting connection to Stellar overlay network...");
+	let mut overlay_conn = connect_to_stellar_overlay_network(config.clone(), secret_key_as_string)
+		.await.map_err(|e|{
+			tracing::error!("listen_for_stellar_messages(): Failed to connect to Stellar overlay network: {e:?}");
+			service::Error::StartOracleAgentError
+		})?;
 
-	let mut overlay_conn = connect_to_stellar_overlay_network(config.clone(), secret_key).await?;
 	// use StellarOverlayConnection's sender to send message to Stellar
 	let sender = overlay_conn.sender();
+	{
+		oracle_agent.write().await.message_sender = Some(sender.clone());
```

Comment:
The whole reason why `Arc<OracleAgent>` has become `Arc<RwLock<OracleAgent>>`. We save the `message_sender` once `overlay_conn` has been initialized.

Reply:
So the reason we cannot call `connect_to_stellar_overlay_network` to get the `overlay_conn` over [here](https://github.com/pendulum-chain/spacewalk/blob/9cad8e4893182ab90cd2483d121df98f048eb402/clients/vault/src/system.rs#L815) and pass the sender there to the `oracle_agent` is because the overlay process will start to receive messages while the `oracle_agent` is not be ready ? Ready I mean not yet started the `read` loop in this function bellow.

Reply:
Yes.

`OracleAgent` is our middleman to use the `stellar-relay-lib`'s connection.  Right now the `overlay_conn` lives in the logic of `OracleAgent`. I do not want to expose `overlay_conn` anywhere else, aside from the `oracle` directory. Separation of concerns.


But I do see your point. If we declare the `overlay_conn` separately, then we can keep `Arc<OracleAgent>`.
I am going to explore your idea; see how cleanly I can do this. 

Thanks Gianni!

Reply:
Okay I see clearly now. No need to change it!

You need to create the oracle agent on the high level [execution](https://github.com/pendulum-chain/spacewalk/blob/9cad8e4893182ab90cd2483d121df98f048eb402/clients/vault/src/system.rs#L782) to pass it to all the other tasks. 
Before, the `overlay_conn` task was created along with the `oracle_agent` [here](https://github.com/pendulum-chain/spacewalk/blob/7c7989875e95e1cfe3b0aeba9bb6af01d9e33e58/clients/vault/src/oracle/agent.rs#L72) and now that second process is spawned inside `listen_for_stellar_messages`. 

I think the separation of concerns is very valid, I just wanted to understand. And I don't see any pro-con regarding where we spawn the processes so far, since we are not handling them (poll_messages_from_stellar) anyway and live independently.